pageserver/src/repository.rs

```diff
@@ -273,14 +281,42 @@ pub trait TimelineWriter: Deref<Target = dyn Timeline> {
     fn advance_last_record_lsn(&self, lsn: Lsn);
 }
 
+/// Each update to a page is represented by a ZenithWalRecord. It can be a wrapper
+/// around a PostgreSQL WAL record, or a custom zenith-specific "record".
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
-pub struct WALRecord {
-    pub will_init: bool,
-    pub rec: Bytes,
-    // Remember the offset of main_data in rec,
-    // so that we don't have to parse the record again.
-    // If record has no main_data, this offset equals rec.len().
-    pub main_data_offset: u32,
+pub enum ZenithWalRecord {
+    /// Native PostgreSQL WAL record
+    Postgres { will_init: bool, rec: Bytes },
+
```

Comment:
I left it as it is for now. I agree separate structs would be nice for passing around a specific variant of the enum, but we don't actually do that currently, So for now, it would just make it more verbose. Let's refactor it later if it becomes more useful.

`walredo.rs` contains functions that take either the Postgres variant or the other variants, and there are currently panics to check that you pass the right kind. Separate structs would seem like a good fit there. But if we wanted to pass a Vec<PostgresWalRecords> for example, then we'd need to make a copy of the Vec which would add some extra code and overhead too.